src/Primitives.hs

```diff
@@ -430,48 +409,36 @@ primitiveMembers _ ctx target = do
 
 -- | Set meta data for a Binder
 primitiveMetaSet :: TernaryPrimitiveCallback
-primitiveMetaSet _ ctx target@(XObj (Sym (SymPath prefixes name) _) _ _) (XObj (Str key) _ _) value =
+primitiveMetaSet _ ctx (XObj (Sym path _) _ _) (XObj (Str key) _ _) value =
   pure $ maybe create (,dynamicNil) lookupAndUpdate
   where
-    fullPath@(SymPath modules _) = consPath (contextPath ctx `union` prefixes) (SymPath [] name)
-    dynamicPath = consPath ["Dynamic"] fullPath
-    global = contextGlobalEnv ctx
-    types = getTypeEnv (contextTypeEnv ctx)
+    qpath = qualifyPath ctx path
+    fullPath@(SymPath modules _) = unqualify qpath
     lookupAndUpdate :: Maybe Context
     lookupAndUpdate =
-      ( lookupBinder dynamicPath global
+      ( lookupBinderInGlobalEnv ctx path
           >>= \binder ->
             pure (Meta.updateBinderMeta binder key value)
-              >>= \b ->
-                pure (envInsertAt global dynamicPath b)
-                  >>= \env -> pure (ctx {contextGlobalEnv = env})
+              >>= pure . (insertInGlobalEnv ctx qpath)
       )
-        <|> ( lookupBinder fullPath global
-                >>= \binder ->
-                  pure (Meta.updateBinderMeta binder key value)
-                    >>= \b ->
-                      pure (envInsertAt global fullPath b)
-                        >>= \env -> pure (ctx {contextGlobalEnv = env})
-            )
         -- This is a global name but it doesn't exist in the global env
         -- Before creating a new binder, check that it doesn't denote an existing type or interface.
         <|> if null modules
           then
-            lookupBinder fullPath types
+            lookupBinderInTypeEnv ctx qpath
               >>= \binder ->
                 pure (Meta.updateBinderMeta binder key value)
-                  >>= \b ->
-                    pure (envInsertAt types fullPath b)
-                      >>= \env -> pure (ctx {contextTypeEnv = TypeEnv env})
+                  >>= pure . (insertInTypeEnv ctx qpath)
           else Nothing
     create :: (Context, Either EvalError XObj)
     create =
-      if null prefixes
-        then
-          let updated = Meta.updateBinderMeta (Meta.stub fullPath) key value
-              newEnv = envInsertAt global fullPath updated
-           in (ctx {contextGlobalEnv = newEnv}, dynamicNil)
-        else evalError ctx ("`meta-set!` failed, I can't find the symbol `" ++ pretty target ++ "`") (xobjInfo target)
+      let updated = Meta.updateBinderMeta (Meta.stub fullPath) key value
+       in (insertInGlobalEnv ctx qpath updated, dynamicNil)
+--if null prefixes
+--  then
+--    let updated = Meta.updateBinderMeta (Meta.stub fullPath) key value
+--     in (insertInGlobalEnv ctx fullPath updated, dynamicNil)
+--  else evalError ctx ("`meta-set!` failed, I can't find the symbol `" ++ pretty target ++ "`") (xobjInfo target)
```

Comment:
Can we get rid of this?

Reply:
Actually, I likely need to restore it. This removes the error for calling `meta-set` against a qualified path so that `(doc Foo.bar "")` would work. Of course, if you actually try to evaluate this, you'll get the "can't insert into non-existing module" error when `Foo` does not yet exist. Likely we should just make that error an actual error since it's more general.

Reply:
Oh yeah, that makes sense of course!

Reply:
restored for now f61526d--I'll add proper error handling to `envInsertAt` in a folllow-up.